### ISSUE TITLE
lex cannot correctly recognize escape sequences.

### ISSUE THREAD

**Issue by Cynthia-Cheng** [sqlfluff/sqlfluff#6064]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When parsing the following SQL, there were false positive lex errors and parse errors.
```
SELECT table_name,
CONCAT('{"name":"', replace(replace(column_name, "\\", "\\\\"), '"', '\\"'), '","comment":"', replace(replace(column_comment, "\\", "\\\\"), '"', '\\"'), '"}') AS json_obj
FROM tb
```
SQLLexError('Unable to lex characters: \'\\\\\\\\"),\'')
SQLLexError('Unable to lex characters: \'\\\\\\\\"),\'')

### Expected Behaviour

As above.

### Observed Behaviour

LexedElement(raw='"\\", "', matcher=<RegexLexer: double_quote>)

The lexer recognized `"\\", "` as a double_quote, indicating that it did not correctly escape them.

### How to reproduce

```
template_str = """
SELECT table_name,
CONCAT('{"name":"', replace(replace(column_name, "\\", "\\\\"), '"', '\\"'), '","comment":"', replace(replace(column_comment, "\\", "\\\\"), '"', '\\"'), '"}') AS json_obj
FROM tb
;
"""
res=sqlfluff.lint(template_str)
```

### Dialect

ansi

### Version

python 3.9
sqlfluff 3.0.7

### Configuration

default

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
With the latest `sqlfluff` the above statement parses properly
